### PR TITLE
Ticket6079 Add Aeroflex IOCs

### PR DESCRIPTION
**Pull request by Scott-James-Hurley**
### Description of work

Adds IOCs for Aeroflex 2023A and 2030 models. Only the functionality required by the ticket was implemented

### To test

Run Aeroflex tests:
https://github.com/ISISComputingGroup/EPICS-IOC_Test_Framework/pull/572

https://github.com/ISISComputingGroup/IBEX/issues/6079

### Acceptance criteria

*List the acceptance criteria for the PR*

---

#### Code Review

- [ ] A copy of the manual has been placed on the shared drive
- [ ] Pertitent information has been stored in the [wiki](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Specific-Device-IOC)
- [ ] Does the IOC conform to IBEX standards?
    - [PV naming](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/PV-Naming).
    - [Disable records](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Disable-records)
    - [Record simulation](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Record-Simulation)
    - [Finishing touches](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/IOC-Finishing-Touches)
- [ ] If an OPI has been modified, does it conform to the [style guidelines](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/OPI-Creation)?
- [ ] Do the IOC and support module conform to the new [build guidelines](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Reducing-Build-Dependencies)
- [ ] Have the changes been recorded appropriately in a PR for [release notes](https://github.com/ISISComputingGroup/IBEX/blob/master/release_notes/ReleaseNotes_Upcoming.md)?
- [ ] Is the device's flow control setting correct? [For most devices flow control should be OFF](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Flow-control).

### Functional Tests

- IOC responds correctly in:
    - [ ] Devsim mode
    - [ ] Recsim mode
    - [ ] Real device, if available
- [ ] Supplementary IOCs (`..._0n` where `n>1`) run correctly
- [ ] Log files do not report undefined macros (serach for `macLib: macro` to find instances of `macLib: macro [macro name] is undefined...`

### Final steps

- [ ] Update the IOC submodule in the main EPICS repo. See [Git workflow](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Git-workflow) page for details.
- [ ] Reviewer has merged the associated PR for the [release notes](https://github.com/ISISComputingGroup/IBEX/blob/master/release_notes/ReleaseNotes_Upcoming.md)

### CODE REVIEW

File: AEROFLEX/iocBoot/iocAEROFLEX-IOC-01/config.xml [hunk 1]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" ?>
+<ioc_config xmlns:xi="http://www.w3.org/2001/XInclude">
+<config_part>
+<ioc_desc>Aeroflex</ioc_desc>
+<ioc_details>Aeroflex</ioc_details>
+<macros>
+<macro name="PORT" pattern="^COM[0-9]+$" description="Serial COM port" hasDefault="NO" />
+<macro name="DEV_TYPE" pattern="^(2023A|2030)$" description="Type of aeroflex; default 2023A" defaultValue="2023A" hasDefault="YES" />
```
Comment (NikolaRoev) [inline AEROFLEX/iocBoot/iocAEROFLEX-IOC-01/config.xml:8]:
```suggestion
<macro name="DEV_TYPE" pattern="^(2023A|2030)$" description="Type of Aeroflex" defaultValue="2023A" hasDefault="YES" />
```
The default value is shown in the client.

**Review by NikolaRoev — CHANGES_REQUESTED** [review]

File: AEROFLEX/iocBoot/iocAEROFLEX-IOC-01/config.xml [hunk 2]
```diff
@@ -6,6 +6,7 @@
 <macros>
 <macro name="PORT" pattern="^COM[0-9]+$" description="Serial COM port" hasDefault="NO" />
 <macro name="DEV_TYPE" pattern="^(2023A|2030)$" description="Type of Aeroflex" defaultValue="2023A" hasDefault="YES" />
+<macro name="RF_PREC" pattern="^(6|2)$" description="RF level decimal precision" defaultValue="6" hasDefault="YES" />
```
Comment (NikolaRoev) [inline AEROFLEX/iocBoot/iocAEROFLEX-IOC-01/config.xml:9]: I can see this being an issue. Let's say I change my device type but forget to adjust the precision. Is something going to break? Ideally the precision should depend on the device type.
Using a `scalcout` record on the `DEV_TYPE` macro should work.

Comment (Scott-James-Hurley) [reply AEROFLEX/iocBoot/iocAEROFLEX-IOC-01/config.xml:9]: I'd like to set the value of `RF_PREC` based on the value of `DEV_TYPE`. Can this be done with `st-common.cmd`?

Comment (NikolaRoev) [reply AEROFLEX/iocBoot/iocAEROFLEX-IOC-01/config.xml:9]: You should be able to use an `if` statement in `st-common.cmd` on the device type macro, and then depending on that result set a "precision" variable and pass that as the precision macro value.

Comment (Scott-James-Hurley) [reply AEROFLEX/iocBoot/iocAEROFLEX-IOC-01/config.xml:9]: I tried adding the following line to `st-common.cmd`:
`if $(DEV_TYPE)==2023A (SET rf_prec = 6) else (SET rf_prec = 2)`
And I get the following error:
`st-common.cmd line 31: Command if not found.`

File: AEROFLEX/iocBoot/iocAEROFLEX-IOC-01/st-common.cmd [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+## Load record instances
+
+##ISIS## Load common DB records 
+< $(IOCSTARTUP)/dbload.cmd
+
+## Load our record instances
+stringiftest("2023A" "$(DEV_TYPE="2023A")" 5 "2023A")
+$(IF2023A) dbLoadRecords("$(AEROFLEX)/db/aeroflex_common.db","PVPREFIX=$(MYPVPREFIX),P=$(MYPVPREFIX)$(IOCNAME):,RECSIM=$(RECSIM=0),DISABLE=$(DISABLE=0),PORT=$(DEVICE),DEV_TYPE=$(DEV_TYPE=2023A),RF_PREC=6")
+$(IFNOT2023A) dbLoadRecords("$(AEROFLEX)/db/aeroflex_common.db","PVPREFIX=$(MYPVPREFIX),P=$(MYPVPREFIX)$(IOCNAME):,RECSIM=$(RECSIM=0),DISABLE=$(DISABLE=0),PORT=$(DEVICE),DEV_TYPE=$(DEV_TYPE=2023A),RF_PREC=2")
```
Comment (NikolaRoev) [inline AEROFLEX/iocBoot/iocAEROFLEX-IOC-01/st-common.cmd:34]:
```suggestion
$(IFNOT2023A) dbLoadRecords("$(AEROFLEX)/db/aeroflex_common.db","PVPREFIX=$(MYPVPREFIX),P=$(MYPVPREFIX)$(IOCNAME):,RECSIM=$(RECSIM=0),DISABLE=$(DISABLE=0),PORT=$(DEVICE),DEV_TYPE=$(DEV_TYPE=2030),RF_PREC=2")
```
Correct device type macro?

File: AEROFLEX/iocBoot/iocAEROFLEX-IOC-01/st-common.cmd [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+dbLoadRecords("$(AEROFLEX)/db/aeroflex_common.db","PVPREFIX=$(MYPVPREFIX),P=$(MYPVPREFIX)$(IOCNAME):,RECSIM=$(RECSIM=0),DISABLE=$(DISABLE=0),PORT=$(DEVICE),DEV_TYPE=$(DEV_TYPE=2023A)")
```
Comment (NikolaRoev) [inline AEROFLEX/iocBoot/iocAEROFLEX-IOC-01/st-common.cmd:36]: This should already be loaded, no?

**Review by NikolaRoev — APPROVED** [review]

**Pull request merged**